src/kernel/src/time.rs

```diff
@@ -0,0 +1,26 @@
+use alloc::{
+    sync::Arc,
+    vec::Vec
+};
+
+use twizzler_abi::syscall::{ClockInfo, FemtoSeconds};
+
+pub struct Ticks {
+  pub value: u64,
+  pub rate: FemtoSeconds
+}
+
+pub trait ClockHardware {
+  fn read(&self) -> Ticks;
+  fn info(&self) -> ClockInfo;
+}
+
+// struct DummySource;
+// impl ClockHardware for DummySource {}
+// register_clock(DummySource); // slots 0 and 1
+
+pub static mut TICK_SOURCES: Vec<Arc<dyn ClockHardware>> = Vec::new();
```

Comment:
You should put a lock around this and make it non-mut, that will remove the need for unsafe (as written, this code violates safety since register_clock could be called concurrently with reading TICK_SOURCES).

It can probably be an rwlock, since updates to TICK_SOURCES is probably rare.

Reply:
So we do not jumble everything into one PR, I plan on implementing a rwlock in a separate PR. An then integrating that with TICK_SOURCES later. For now, TICK_SOURCES is protected using a spinlock.